### DISCUSSION TITLE
Segregate key bindings per operation mode

### DISCUSSION THREAD

**Discussion by volodymyrprokopyuk** [fish-shell/fish-shell#10083] [Q&A] [answered]
Hi,

Is it possible to `bind` the same key combination to different functions in different modes e. g. accept auto suggestion and select from a tab completion?

I have `bind \em accept-autosuggestion` to accept an auto suggestion with `Alt-m`.

I want to `bind \em down-line` and `bind \e, up-line` to trigger only in the tab completion mode to select a completion with `Alt-m` to move down and `Alt-,` to move up in a completion drop-down list.

The problem with the above configuration is that `Alt-m` is always bound to `down-line`. Is it possible to segregate key bindings per operation mode?

Thank you!

**Comment by faho** [comment, marked as answer]
You want `commandline --paging-mode`, like


```fish
bind \em 'if commandline --paging-mode; commandline -f down-line; else; ...; end'
```

You can also make this a function and bind that, see e.g. [down-or-search](https://github.com/fish-shell/fish-shell/blob/master/share/functions/down-or-search.fish).

**Reply by volodymyrprokopyuk** [reply]
Fantastic! Thank you very much! I really enjoy fish! It is very convenient and it just works the expected way!

Now I have

```fish
bind \em 'if commandline --paging-mode; commandline -f down-line; \
  else; commandline -f accept-autosuggestion; end'
bind \e, up-line
```

**Comment by volodymyrprokopyuk** [comment]
HI,

A related question that I struggle to find a solution. Is it possible to bind `Ctrl-r` in the insert mode of the vi mode?

I have the below configuration that works in the normal mode of the vi mode, but does not work in the insert mode. It is quite convenient to search the command history directly from the insert mode without having to switch to the normal mode.

```fish
function fzf_history
  history --null | fzf --query=(commandline) --read0 --print0 \
    --cycle --bind=alt-m:down,alt-,:up --ansi | \
    read --null --local result; and commandline -- $result
end
bind \cR fzf_history
```

Following the above suggested pattern there is no option like `commandline --insert-mode`.

**Reply by faho** [reply]
`bind -M insert yourkey yourfunction`

**Reply by volodymyrprokopyuk** [reply]
Thank you!
